api/config_test.go

```diff
@@ -0,0 +1,296 @@
+package api_test
+
+import (
+	"io/ioutil"
+	"os"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	. "github.com/starkandwayne/shield/api"
+)
+
+var _ = Describe("API Config", func() {
+	Describe("When loading configs", func() {
+		defaultCfg := &Config{Backends: map[string]string{}, Aliases: map[string]string{}}
+		BeforeEach(func() {
+			os.Chmod("test/etc/unreadable.yml", 0200)
+		})
+		AfterEach(func() {
+			os.Chmod("test/etc/unreadable.yml", 0644)
+		})
+		It("Throws an error on invalid yaml files", func() {
+			Expect(LoadConfig("test/etc/invalid.yml")).ShouldNot(Succeed())
+			defaultCfg.Path = "test/etc/invalid.yml"
+			Expect(Cfg).Should(Equal(defaultCfg))
+		})
+		It("Throws an error on unreadable files", func() {
+			Expect(LoadConfig("test/etc/unreadable.yml")).ShouldNot(Succeed())
+			defaultCfg.Path = "test/etc/unreadable.yml"
+			Expect(Cfg).Should(Equal(defaultCfg))
+		})
+		It("Succeeds if no config was found", func() {
+			Expect(LoadConfig("test/etc/missing.yml")).Should(Succeed())
+			defaultCfg.Path = "test/etc/missing.yml"
+			Expect(Cfg).Should(Equal(defaultCfg))
+		})
+		It("Reads configs and sets up the api.Cfg variable if config was valid", func() {
+			Expect(LoadConfig("test/etc/valid.yml")).Should(Succeed())
+
+			valid := &Config{
+				Backends: map[string]string{
+					"http://first":  "basic mytoken1",
+					"http://second": "basic mytoken2",
+				},
+				Aliases: map[string]string{
+					"first":  "http://first",
+					"second": "http://second",
+				},
+				Backend: "first",
+				Path:    "test/etc/valid.yml",
+			}
+			Expect(Cfg).Should(Equal(valid))
+		})
+	})
+	Describe("When saving configs", func() {
+		It("Throws an error when failing to write data", func() {
+			cfg := &Config{Backend: "default", Path: "/path/to/nowhere"}
+			Expect(cfg.Save()).ShouldNot(Succeed())
+		})
+		It("Successfully writes the config to disk", func() {
+			tempFile, err := ioutil.TempFile("", "shield-test-cfg") // get default tmpdir for OS + supply a prefix
+			Expect(err).ShouldNot(HaveOccurred())
+			tempFile.Close()
+			cfg := &Config{Backend: "default", Path: tempFile.Name()}
+			expectedCfg := `backend: default
+backends: {}
+aliases: {}
+`
+
+			Expect(cfg.Save()).Should(Succeed())
+
+			data, err := ioutil.ReadFile(tempFile.Name())
+			Expect(err).ShouldNot(HaveOccurred())
+			Expect(string(data)).Should(Equal(expectedCfg))
+
+			Expect(os.Remove(tempFile.Name())).Should(Succeed())
```

Comment:
Do we really want the tests to fail if there is a problem with removing the tempfile used for testing?  It's not really a reflection of the code under test...